File: src/agents/concept_extractor.py

import json
import logging
import re
from typing import List, Dict, Any, Optional
from uuid import uuid4

from anthropic import Anthropic
from tenacity import (
    retry,
    stop_after_attempt,
    wait_exponential,
    retry_if_exception_type
)

from ..models import (
    ExtractedEntity,
    EntityType,
    StructuredPaper,
    Section,
    Location
)

logger = logging.getLogger(__name__)


class ConceptExtractor:
    """
    Agent for extracting research concepts from academic papers.

    Attributes:
        client: Anthropic API client
        model: Claude model to use (sonnet for complex reasoning)
        max_tokens: Maximum tokens in response
        temperature: Sampling temperature (low for consistency)
    """

    MODEL = "claude-sonnet-4-20250514"
    MAX_TOKENS = 8192
    TEMPERATURE = 0.2  # Low temperature for consistency

    # Prompt template with few-shot examples
    EXTRACTION_PROMPT = """You are a research concept extraction specialist. Analyze this paper section and extract key conceptual terms.

Guidelines:
1. Extract concepts at multiple granularities:
   - Broad: "computer vision", "3D reconstruction"
   - Specific: "neural scene representation", "volumetric ray marching"
2. Focus on concepts central to the paper's contribution
3. Exclude generic terms like "experiment", "result", "paper", "approach", "method" (unless part of a specific named concept)
4. For each concept, provide:
   - Normalized form (e.g., "NeRF" → "Neural Radiance Fields")
   - Aliases (common variations and abbreviations)
   - First mention location (section, paragraph)
   - Importance score (1-10, where 10 is central to paper's contribution)
   - Confidence score (0-1, based on clarity of definition/usage)
   - Definition (if explicitly defined in the text)
   - Category: technique, problem, application, theory

Paper abstract (for context):
{abstract}

Section content:
{section_content}

Output format: JSON array of concepts

Few-shot examples:

Example 1:
Text: "Neural Radiance Fields (NeRF) represent scenes as continuous volumetric functions encoded in neural network weights..."
Output: {{
  "name": "Neural Radiance Fields",
  "normalized_name": "Neural Radiance Fields",
  "aliases": ["NeRF", "NeRFs", "neural radiance field"],
  "importance": 10,
  "confidence": 0.98,
  "first_mention": {{"section": "Introduction", "paragraph": 2}},
  "definition": "Continuous volumetric scene representations encoded in neural network weights",
  "category": "technique"
}}

Example 2:
Text: "We address the view synthesis problem, generating novel viewpoints from a set of input images..."
Output: {{
  "name": "view synthesis",
  "normalized_name": "View Synthesis",
  "aliases": ["novel view synthesis", "image synthesis"],
  "importance": 8,
  "confidence": 0.92,
  "first_mention": {{"section": "Introduction", "paragraph": 1}},
  "definition": "Generating new camera viewpoints from input images",
  "category": "problem"
}}

Example 3:
Text: "Real-time rendering remains a challenge for neural scene representations..."
Output: {{
  "name": "real-time rendering",
  "normalized_name": "Real-Time Rendering",
  "aliases": ["realtime rendering", "interactive rendering"],
  "importance": 9,
  "confidence": 0.95,
  "first_mention": {{"section": "Introduction", "paragraph": 3}},
  "category": "problem"
}}

Now extract concepts from the provided section. Return ONLY a JSON array, no additional text.
"""

    def __init__(self, api_key: str):
        """
        Initialize the Concept Extractor.

        Args:
            api_key: Anthropic API key
        """
        self.client = Anthropic(api_key=api_key)

    async def extract(self, paper: StructuredPaper) -> List[ExtractedEntity]:
        """
        Extract concepts from a structured paper.

        Args:
            paper: Structured paper to extract from

        Returns:
            List of extracted concept entities

        Raises:
            Exception: If extraction fails after retries
        """
        logger.info(f"Starting concept extraction for paper {paper.metadata.arxiv_id}")

        concepts: List[ExtractedEntity] = []

        # Target sections most likely to contain conceptual information
        target_sections = [
            s for s in paper.sections
            if re.search(r"abstract|introduction|related work|background|method",
                        s.title, re.IGNORECASE)
        ]

        if not target_sections:
            logger.warning(
                f"No target sections found for {paper.metadata.arxiv_id}, "
                f"using first 3 sections as fallback"
            )
            target_sections = paper.sections[:3]

        # Extract from each section
        for section in target_sections:
            try:
                section_concepts = await self._extract_from_section(
                    section,
                    paper.metadata.abstract
                )
                concepts.extend(section_concepts)
                logger.debug(
                    f"Extracted {len(section_concepts)} concepts from "
                    f"section '{section.title}'"
                )
            except Exception as e:
                logger.error(
                    f"Failed to extract from section '{section.title}': {e}",
                    exc_info=True
                )
                # Continue with other sections

        # Deduplicate concepts
        deduplicated = self._deduplicate_concepts(concepts)

        logger.info(
            f"Concept extraction completed for {paper.metadata.arxiv_id}: "
            f"{len(concepts)} raw -> {len(deduplicated)} after deduplication"
        )

        return deduplicated

    @retry(
        stop=stop_after_attempt(3),
        wait=wait_exponential(multiplier=1, min=4, max=60),
        retry=retry_if_exception_type((TimeoutError, ConnectionError))
    )
    async def _extract_from_section(
        self,
        section: Section,
        abstract: str
    ) -> List[ExtractedEntity]:
        """
        Extract concepts from a single section with retry logic.

        Args:
            section: Section to extract from
            abstract: Paper abstract for context

        Returns:
            List of extracted entities
        """
        prompt = self._build_prompt(section.content, abstract)

        # Call Claude API
        response = self.client.messages.create(
            model=self.MODEL,
            max_tokens=self.MAX_TOKENS,
            temperature=self.TEMPERATURE,
            messages=[{
                "role": "user",
                "content": prompt
            }]
        )

        # Parse response
        text_content = response.content[0]
        if text_content.type != "text":
            raise ValueError("Unexpected response type from Claude")

        # Extract JSON (may be wrapped in markdown code blocks)
        json_text = self._extract_json(text_content.text)

        try:
            extracted = json.loads(json_text)
        except json.JSONDecodeError as e:
            logger.error(
                f"Failed to parse JSON response: {e}\n"
                f"Response preview: {text_content.text[:500]}"
            )
            return []

        # Convert to ExtractedEntity objects
        return [
            self._to_extracted_entity(concept, section)
            for concept in extracted
        ]

    def _build_prompt(self, section_content: str, abstract: str) -> str:
        """Build the extraction prompt."""
        # Truncate if too long
        max_section_length = 4000
        max_abstract_length = 500

        if len(section_content) > max_section_length:
            section_content = section_content[:max_section_length] + "..."

        if len(abstract) > max_abstract_length:
            abstract = abstract[:max_abstract_length] + "..."

        return self.EXTRACTION_PROMPT.format(
            abstract=abstract,
            section_content=section_content
        )

    def _extract_json(self, text: str) -> str:
        """
        Extract JSON from text that may be wrapped in markdown code blocks.

        Args:
            text: Text potentially containing JSON

        Returns:
            Extracted JSON string
        """
        # Try to find JSON in markdown code block
        json_match = re.search(r'```json\n([\s\S]*?)\n```', text)
        if json_match:
            return json_match.group(1)

        # Try to find plain code block
        code_match = re.search(r'```\n([\s\S]*?)\n```', text)
        if code_match:
            return code_match.group(1)

        # Assume entire text is JSON
        return text.strip()

    def _to_extracted_entity(
        self,
        concept: Dict[str, Any],
        section: Section
    ) -> ExtractedEntity:
        """
        Convert API output to ExtractedEntity model.

        Args:
            concept: Raw concept dict from API
            section: Section it was extracted from

        Returns:
            ExtractedEntity instance
        """
        first_mention = concept.get("first_mention", {})

        return ExtractedEntity(
            id=str(uuid4()),
            type=EntityType.CONCEPT,
            name=concept["name"],
            normalized_name=concept["normalized_name"],
            aliases=concept.get("aliases", []),
            confidence=float(concept["confidence"]),
            importance=int(concept["importance"]),
            first_mention=Location(
                section=first_mention.get("section", section.title),
                paragraph=first_mention.get("paragraph", 0)
            ),
            all_mentions=[Location(
                section=first_mention.get("section", section.title),
                paragraph=first_mention.get("paragraph", 0)
            )],
            definition=concept.get("definition"),
            category=concept.get("category", "technique")
        )

    def _deduplicate_concepts(
        self,
        concepts: List[ExtractedEntity]
    ) -> List[ExtractedEntity]:
        """
        Deduplicate concepts by normalizing names and merging similar ones.

        Args:
            concepts: List of extracted concepts

        Returns:
            Deduplicated list
        """
        # Group by normalized name (case-insensitive)
        grouped: Dict[str, List[ExtractedEntity]] = {}

        for concept in concepts:
            key = concept.normalized_name.lower()
            if key not in grouped:
                grouped[key] = []
            grouped[key].append(concept)

        # Merge duplicates
        deduplicated: List[ExtractedEntity] = []

        for key, group in grouped.items():
            if len(group) == 1:
                deduplicated.append(group[0])
                continue

            # Merge: take highest confidence, combine aliases and mentions
            best = max(group, key=lambda c: c.confidence)

            # Combine all aliases
            all_aliases = set()
            for concept in group:
                all_aliases.add(concept.name)
                all_aliases.update(concept.aliases)

            # Remove canonical name from aliases
            all_aliases.discard(best.normalized_name)
            best.aliases = sorted(all_aliases)

            # Combine all mentions
            best.all_mentions = [
                mention
                for concept in group
                for mention in concept.all_mentions
            ]

            # Take first available definition
            if not best.definition:
                for concept in group:
                    if concept.definition:
                        best.definition = concept.definition
                        break

            deduplicated.append(best)

            logger.debug(
                f"Merged {len(group)} duplicate concepts into "
                f"'{best.normalized_name}' with {len(best.aliases)} aliases"
            )

        return deduplicated

    def validate_concept(
        self,
        concept: ExtractedEntity,
        paper_text: str
    ) -> Dict[str, Any]:
        """
        Validate an extracted concept against the paper text.

        Args:
            concept: Concept to validate
            paper_text: Full text of the paper

        Returns:
            Dictionary with 'valid' (bool) and 'issues' (list) keys
        """
        issues = []
        paper_lower = paper_text.lower()

        # Check if concept appears in paper
        variants = [concept.name, concept.normalized_name] + concept.aliases
        mentioned = any(
            variant.lower() in paper_lower
            for variant in variants
        )

        if not mentioned:
            issues.append("Concept not found in paper text")

        # Check confidence threshold
        if concept.confidence < 0.5:
            issues.append("Confidence below acceptable threshold (0.5)")

        # Check for potential hallucination
        if len(concept.all_mentions) == 1 and concept.importance > 7:
            issues.append(
                "High importance (>7) but mentioned only once "
                "(possible hallucination)"
            )

        return {
            "valid": len(issues) == 0,
            "issues": issues
        }


def create_concept_extractor(api_key: str) -> ConceptExtractor:
    """
    Factory function to create a ConceptExtractor.

    Args:
        api_key: Anthropic API key

    Returns:
        Configured ConceptExtractor instance
    """
    return ConceptExtractor(api_key)


"""
Concept Extractor Agent

Specialized LLM agent for extracting research concepts from academic papers.
Uses Claude Sonnet 4.5 for complex reasoning about conceptual relationships.

This agent:
- Identifies high-level research concepts at multiple granularities
- Normalizes concept names and tracks aliases
- Assesses the importance and confidence for each extraction
- Validates extractions against paper text
"""

File: src/agents/ConceptExtractor.ts

import Anthropic from '@anthropic-ai/sdk';
import { v4 as uuid } from 'uuid';
import {
  ExtractedEntity,
  EntityType,
  StructuredPaper,
  Section,
  Location
} from '../types';
import { logger } from '../utils/logger';
import { withRetry } from '../utils/retry';
import { ErrorType } from '../types';

interface ConceptExtractionOutput {
  name: string;
  normalized_name: string;
  aliases: string[];
  importance: number;
  confidence: number;
  first_mention: {
    section: string;
    paragraph: number;
  };
  definition?: string;
  category?: 'technique' | 'problem' | 'metric' | 'dataset' | 'application' | 'theory';
}

export class ConceptExtractor {
  private client: Anthropic;
  private readonly model = 'claude-sonnet-4-20250514';
  private readonly maxTokens = 8192;
  private readonly temperature = 0.2; // Low temperature for consistency

  constructor(apiKey: string) {
    this.client = new Anthropic({ apiKey });
  }

  /**
   * Extract concepts from a structured paper
   */
  async extract(paper: StructuredPaper): Promise<ExtractedEntity[]> {
    const span = logger.startSpan('concept_extractor.extract', {
      paperId: paper.metadata.arxivId
    });

    try {
      const concepts: ExtractedEntity[] = [];

      // Focus on sections most likely to contain conceptual information
      const targetSections = paper.sections.filter(s =>
        /abstract|introduction|related work|background|method/i.test(s.title)
      );

      if (targetSections.length === 0) {
        logger.warn('No target sections found for concept extraction', {
          paperId: paper.metadata.arxivId,
          availableSections: paper.sections.map(s => s.title)
        });
        // Fallback: use first 3 sections
        targetSections.push(...paper.sections.slice(0, 3));
      }

      // Extract from each section
      for (const section of targetSections) {
        try {
          const sectionConcepts = await this.extractFromSection(
            section,
            paper.metadata.abstract
          );
          concepts.push(...sectionConcepts);
        } catch (error) {
          logger.error('Failed to extract concepts from section', {
            paperId: paper.metadata.arxivId,
            section: section.title,
            error: error.message
          });
          // Continue with other sections
        }
      }

      // Deduplicate concepts
      const deduplicated = this.deduplicateConcepts(concepts);

      logger.info('Concept extraction completed', {
        paperId: paper.metadata.arxivId,
        sectionsProcessed: targetSections.length,
        conceptsExtracted: concepts.length,
        afterDeduplication: deduplicated.length
      });

      span.setAttribute('concepts_extracted', deduplicated.length);
      return deduplicated;

    } catch (error) {
      logger.error('Concept extraction failed', {
        paperId: paper.metadata.arxivId,
        error: error.message
      });
      throw error;
    } finally {
      span.end();
    }
  }

  /**
   * Extract concepts from a single section
   */
  private async extractFromSection(
    section: Section,
    abstract: string
  ): Promise<ExtractedEntity[]> {

    const prompt = this.buildPrompt(section.content, abstract);

    const response = await withRetry(
      async () => {
        return await this.client.messages.create({
          model: this.model,
          max_tokens: this.maxTokens,
          temperature: this.temperature,
          messages: [{
            role: 'user',
            content: prompt
          }]
        });
      },
      ErrorType.API_TIMEOUT,
      { section: section.title }
    );

    // Parse response
    const textContent = response.content[0];
    if (textContent.type !== 'text') {
      throw new Error('Unexpected response type from Claude');
    }

    let extracted: ConceptExtractionOutput[];
    try {
      // Extract JSON from response (may be wrapped in markdown code blocks)
      const jsonMatch = textContent.text.match(/```json\n([\s\S]*?)\n```/);
      const jsonText = jsonMatch ? jsonMatch[1] : textContent.text;
      extracted = JSON.parse(jsonText);
    } catch (error) {
      logger.error('Failed to parse concept extraction response', {
        section: section.title,
        response: textContent.text.substring(0, 500),
        error: error.message
      });
      return []; // Return empty array rather than failing entire extraction
    }

    // Convert to ExtractedEntity format
    return extracted.map(concept => this.toExtractedEntity(concept, section));
  }

  /**
   * Build the extraction prompt with few-shot examples
   */
  private buildPrompt(sectionContent: string, abstract: string): string {
    return `You are a research concept extraction specialist. Analyze this paper section and extract key conceptual terms.

Guidelines:
1. Extract concepts at multiple granularities:
   - Broad: "computer vision", "3D reconstruction"
   - Specific: "neural scene representation", "volumetric ray marching"
2. Focus on concepts central to the paper's contribution
3. Exclude generic terms like "experiment", "result", "paper", "approach", "method" (unless part of a specific named concept)
4. For each concept, provide:
   - Normalized form (e.g., "NeRF" → "Neural Radiance Fields")
   - Aliases (common variations and abbreviations)
   - First mention location (section, paragraph)
   - Importance score (1-10, where 10 is central to paper's contribution)
   - Confidence score (0-1, based on clarity of definition/usage)
   - Definition (if explicitly defined in the text)
   - Category: technique, problem, application, theory

Paper abstract (for context):
${abstract.substring(0, 500)}...

Section content:
${sectionContent}

Output format: JSON array of concepts

Few-shot examples:

Example 1:
Text: "Neural Radiance Fields (NeRF) represent scenes as continuous volumetric functions encoded in neural network weights..."
Output: {
  "name": "Neural Radiance Fields",
  "normalized_name": "Neural Radiance Fields",
  "aliases": ["NeRF", "NeRFs", "neural radiance field"],
  "importance": 10,
  "confidence": 0.98,
  "first_mention": {"section": "Introduction", "paragraph": 2},
  "definition": "Continuous volumetric scene representations encoded in neural network weights",
  "category": "technique"
}

Example 2:
Text: "We address the view synthesis problem, generating novel viewpoints from a set of input images..."
Output: {
  "name": "view synthesis",
  "normalized_name": "View Synthesis",
  "aliases": ["novel view synthesis", "image synthesis"],
  "importance": 8,
  "confidence": 0.92,
  "first_mention": {"section": "Introduction", "paragraph": 1},
  "definition": "Generating new camera viewpoints from input images",
  "category": "problem"
}

Example 3:
Text: "Real-time rendering remains a challenge for neural scene representations..."
Output: {
  "name": "real-time rendering",
  "normalized_name": "Real-Time Rendering",
  "aliases": ["realtime rendering", "interactive rendering"],
  "importance": 9,
  "confidence": 0.95,
  "first_mention": {"section": "Introduction", "paragraph": 3},
  "category": "problem"
}

Now extract concepts from the provided section. Return ONLY the JSON array, no additional text.`;
  }

  /**
   * Convert API output to ExtractedEntity
   */
  private toExtractedEntity(
    concept: ConceptExtractionOutput,
    section: Section
  ): ExtractedEntity {
    return {
      id: uuid(),
      type: EntityType.CONCEPT,
      name: concept.name,
      normalizedName: concept.normalized_name,
      aliases: concept.aliases || [],
      confidence: concept.confidence,
      importance: concept.importance,
      firstMention: {
        section: concept.first_mention.section,
        paragraph: concept.first_mention.paragraph
      },
      allMentions: [{
        section: concept.first_mention.section,
        paragraph: concept.first_mention.paragraph
      }],
      definition: concept.definition,
      category: concept.category || 'technique'
    };
  }

  /**
   * Deduplicate concepts by normalizing names and merging similar ones
   */
  private deduplicateConcepts(concepts: ExtractedEntity[]): ExtractedEntity[] {
    const grouped = new Map<string, ExtractedEntity[]>();

    // Group by normalized name (case-insensitive)
    for (const concept of concepts) {
      const key = concept.normalizedName.toLowerCase();
      if (!grouped.has(key)) {
        grouped.set(key, []);
      }
      grouped.get(key)!.push(concept);
    }

    // Merge duplicates
    const deduplicated: ExtractedEntity[] = [];
    for (const [key, group] of grouped.entries()) {
      if (group.length === 1) {
        deduplicated.push(group[0]);
        continue;
      }

      // Merge: take highest confidence, combine aliases and mentions
      const merged = group.reduce((best, current) => {
        if (current.confidence > best.confidence) {
          return current;
        }
        return best;
      });

      // Combine aliases from all duplicates
      const allAliases = new Set<string>();
      for (const concept of group) {
        allAliases.add(concept.name);
        concept.aliases.forEach(alias => allAliases.add(alias));
      }
      allAliases.delete(merged.normalizedName); // Remove canonical name from aliases
      merged.aliases = Array.from(allAliases);

      // Combine all mentions
      merged.allMentions = group.flatMap(c => c.allMentions);

      // Take definition from any that has one
      if (!merged.definition) {
        merged.definition = group.find(c => c.definition)?.definition;
      }

      deduplicated.push(merged);

      logger.debug('Merged duplicate concepts', {
        normalizedName: merged.normalizedName,
        duplicateCount: group.length,
        aliases: merged.aliases
      });
    }

    return deduplicated;
  }

  /**
   * Validate extracted concept
   */
  validateConcept(
    concept: ExtractedEntity,
    paperText: string
  ): { valid: boolean; issues: string[] } {
    const issues: string[] = [];

    // Check if concept appears in paper
    const variants = [concept.name, concept.normalizedName, ...concept.aliases];
    const mentioned = variants.some(v =>
      paperText.toLowerCase().includes(v.toLowerCase())
    );

    if (!mentioned) {
      issues.push('Concept not found in paper text');
    }

    // Check confidence threshold
    if (concept.confidence < 0.5) {
      issues.push('Confidence below acceptable threshold');
    }

    // Check importance for single mentions
    if (concept.allMentions.length === 1 && concept.importance > 7) {
      issues.push('High importance but mentioned only once (possible hallucination)');
    }

    return {
      valid: issues.length === 0,
      issues
    };
  }
}

/**
 * Factory function to create ConceptExtractor with configuration
 */
export function createConceptExtractor(config: {
  apiKey: string;
}): ConceptExtractor {
  return new ConceptExtractor(config.apiKey);
}
